### PR TITLE
[WIP]Roli/reduceprod2

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by HectorSVC** [comment]
>   Tensor* Y = ctx->Output(0, TensorShape(squeezed_output_dims));

why not remove the duplicate code here, and reuse your PrepareForReduce?

---
Refers to: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:156 in e4fbd26. [](commit_id = e4fbd2686e3a04f22126e3b56aeb11cb6afa7f6f, deletion_comment = False)

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc [hunk 1]
```diff
@@ -259,5 +391,16 @@ REGISTER_KERNEL_HFD(ReduceLogSum)
...
+#define REGISTER_KERNEL_INT32(name) \
+  REGISTER_KERNEL_TYPED(name, int32_t)
+
+REGISTER_KERNEL_INT32(ReduceL1)
+REGISTER_KERNEL_INT32(ReduceL2)
+REGISTER_KERNEL_INT32(ReduceMax)
+REGISTER_KERNEL_INT32(ReduceMean)
+REGISTER_KERNEL_INT32(ReduceMin)
+REGISTER_KERNEL_INT32(ReduceProd)
+REGISTER_KERNEL_INT32(ReduceSum)
```
Comment (HectorSVC) [inline onnxruntime/core/providers/cuda/reduction/reduction_ops.cc]: what about other ops like ArgMax, ArgMin, ReduceLogSum, ReduceSumSquare, ReduceLogSumExp

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc [hunk 2]
```diff
@@ -241,6 +302,77 @@ Status ReduceKernel<allow_multi_axes>::ComputeImpl(OpKernelContext* ctx, cudnnRe
   return Status::OK();
 }
 
+template <>
+template <>
```
Comment (ybrnathan) [inline onnxruntime/core/providers/cuda/reduction/reduction_ops.cc]: you might want to add comments on why you need to handle int32 explicitly

**Comment by skottmckay** [comment]
>   Tensor* Y = ctx->Output(0, TensorShape(squeezed_output_dims));

If you don't do that, you also need to add handling for a dim value of 0. See [https://github.com/microsoft/onnxruntime/pull/2217](https://github.com/microsoft/onnxruntime/pull/2217)

---
In reply to: [544631995](https://github.com/microsoft/onnxruntime/pull/2108#issuecomment-544631995) [](ancestors = 544631995)

---
Refers to: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc:156 in e4fbd26. [](commit_id = e4fbd2686e3a04f22126e3b56aeb11cb6afa7f6f, deletion_comment = False)

File: onnxruntime/core/providers/cuda/reduction/reduction_ops.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cuda/reduction/reduction_ops.cc]: also why are there two 'template<>' lines?

---
In reply to: [338331298](https://github.com/microsoft/onnxruntime/pull/2108#discussion_r338331298) [](ancestors = 338331298)

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 3]
```diff
@@ -76,10 +76,9 @@ CUDAExecutionProvider::CUDAExecutionProvider(const CUDAExecutionProviderInfo& in
 
   // TODO: this is actually used for the cuda kernels which explicitly ask for inputs from CPU.
   // This will be refactored/removed when allocator and execution provider are decoupled.
-  DeviceAllocatorRegistrationInfo cpu_memory_info({
-      OrtMemTypeCPUInput,
-      [](int device_id) { return onnxruntime::make_unique<CPUAllocator>(onnxruntime::make_unique<OrtMemoryInfo>("CUDA_CPU", OrtAllocatorType::OrtDeviceAllocator, OrtDevice(), device_id, OrtMemTypeCPUInput)); },
-      std::numeric_limits<size_t>::max()});
+  DeviceAllocatorRegistrationInfo cpu_memory_info({OrtMemTypeCPUInput,
+                                                   [](int device_id) { return onnxruntime::make_unique<CPUAllocator>(onnxruntime::make_unique<OrtMemoryInfo>("CUDA_CPU", OrtAllocatorType::OrtDeviceAllocator, OrtDevice(), device_id, OrtMemTypeCPUInput)); },
+                                                   std::numeric_limits<size_t>::max()});
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: nit: line is way too long. max is 120 chars

**Comment by yufenglee** [comment]
This PR is not ready for review. Need more work on it.

**Pull request closed without merging**